source/slang/slang-check-decl.cpp

```diff
@@ -9308,8 +9308,272 @@ namespace Slang
+            {
+                // Forward declared constructors may be declared in an interface, these must be ignored.
+                // We can determine these based on the body being nullptr
+                auto ctor = as<ConstructorDecl>(m);
+                if (!ctor || !ctor->body)
+                    continue;
+                aggTypeInfo->ctorList.add(ctor);
+                if (ctor->members.getCount() != 0)
+                    continue;
+                aggTypeInfo->defCtor = ctor;
+            }
+        }
+        return inheritanceDepth;
+    }
+
+    static ConstructorDecl* _createCtor(SemanticsDeclAttributesVisitor* visitor, ASTBuilder* m_astBuilder, AggTypeDecl* decl)
+    {
+        auto ctor = m_astBuilder->create<ConstructorDecl>();
+        auto ctorName = visitor->getName("$init");
+        ctor->ownedScope = m_astBuilder->create<Scope>();
+        ctor->ownedScope->containerDecl = ctor;
+        ctor->ownedScope->parent = visitor->getScope(decl);
+        ctor->parentDecl = decl;
+        ctor->loc = decl->loc;
+        ctor->closingSourceLoc = ctor->loc;
+        ctor->nameAndLoc.name = ctorName;
+        ctor->nameAndLoc.loc = ctor->loc;
+        ctor->returnType.type = visitor->calcThisType(makeDeclRef(decl));
+        auto body = m_astBuilder->create<BlockStmt>();
+        body->scopeDecl = m_astBuilder->create<ScopeDecl>();
+        body->scopeDecl->ownedScope = m_astBuilder->create<Scope>();
+        body->scopeDecl->ownedScope->parent = visitor->getScope(ctor);
+        body->scopeDecl->parentDecl = ctor;
+        body->scopeDecl->loc = ctor->loc;
+        body->scopeDecl->closingSourceLoc = ctor->loc;
+        body->closingSourceLoc = ctor->closingSourceLoc;
+        ctor->body = body;
+        body->body = m_astBuilder->create<SeqStmt>();
+        decl->addMember(ctor);
+        return ctor;
+    }
+
+    static SeqStmt* legalizeCtorBlockToASeqStmt(ASTBuilder* m_astBuilder, BlockStmt* stmt)
+    {
+        // It is possible BlockStmt has a child with the type of
+        // `ExpressionStmt` if an existing constructor has only 1
+        // expression. This would be a senario we need to
+        // put the `ExpressionStmt` inside a `SeqStmt`.
+        // We need to ensure we are adding to a constructor which can
+        // handle multiple statments since we may be adding field `init`
+        // expressions and function calls to the constructor.
+        if (!as<SeqStmt>(stmt->body))
+        {
+            auto tmpExpr = stmt->body;
+            auto seqStmt = m_astBuilder->create<SeqStmt>();
+            seqStmt->stmts.add(tmpExpr);
+            stmt->body = seqStmt;
+            return seqStmt;
+        }
+        return as<SeqStmt>(stmt->body);
+    }
+    static void legalizeCtorOverloadRank(ASTBuilder* m_astBuilder, ConstructorDecl* ctor, Facet::DirectnessVal rank)
+    {
+        // It is possible a "this" has a "super". Since we are creating a new constructor
```

Comment:
Instead of hacking this with overloadRank, we should implement the proper logic in resolveOverload to prefer the method from a more derived type. Note that this is necessary not only for constructors but also other member methods.